### PR TITLE
FEATURE: Add KaTeX in-browser rendering support.

### PR DESCRIPTION
**Pull request by misaka4e21**

### CODE REVIEW

File: assets/javascripts/initializers/discourse-math-katex.js.es6 [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+  if ($elem.data("applied-katex")) {
+    return;
+  }
+  $elem.data("applied-katex", true);
+
+  if ($elem.hasClass("math")) {
+    const text = $elem.text();
+    $elem.text('');
+    window.katex.render(text, elem, {
```
Comment (ZogStriP) [inline assets/javascripts/initializers/discourse-math-katex.js.es6]: Using ES6, you can shorten this a bit

```javascript
window.katex.render(text, elem, { displayMode })
```

File: assets/javascripts/initializers/discourse-math-katex.js.es6 [hunk 2]
```diff
@@ -0,0 +1,60 @@
+import { withPluginApi } from "discourse/lib/plugin-api";
+import loadScript from "discourse/lib/load-script";
+
+function ensureKaTeX() {
+  return loadScript("/plugins/discourse-math/katex/katex.min.js").then(() => {
```
Comment (danielhollas) [inline assets/javascripts/initializers/discourse-math-katex.js.es6]: Sorry for probably a stupid question, but shouldn't these be included in the repo?

Comment (misaka4e21) [reply assets/javascripts/initializers/discourse-math-katex.js.es6]: Fixed. Feeling sorry that I forget to do `git add` on these files.

**Comment by danielhollas** [comment]
Apparently, it could be trimmed in exchange for lack of older browser support.
https://khan.github.io/KaTeX/docs/font.html#kinds-of-fonts-used

**Comment by SamSaffron** [comment]
woff and woff2 only is fine, Discourse does not work in any browsers
earlier than that anyway. My bigger question though are any of the 20 or so
fonts superfluous?

**Comment by danielhollas** [comment]
I dunno tbh, I don't see any obvious candidates. I'd guess that as long as you want to support all the features, you should keep them. Philosophy of Katex is to be fast and small, so I'd guess they wouldn't add superfluous stuff. But again, this is a guesswork on my part.

**Comment by danielhollas** [comment]
PS: They've just released version 0.10.0-rc, so maybe we could wait for stable version 0.10.0?

**Comment by danielhollas** [comment]
FYI: @SamSaffron KaTeX just released a new stable version 0.10
https://github.com/Khan/KaTeX/releases/tag/v0.10.0

It would be great if this could be finished. :-) There's a still the problem with loading fonts described above. I am happy to help with testing, but do not know enough about rails or other stuff to fix it. :-(

**Comment by SamSaffron** [comment]
I want to get to this but we have so many other projects going at the moment, can you ping me again to remind me in 3-4 weeks?

**Comment by cycomachead** [comment]
Is there any bandwidth to work on this soon? I'm on a project that uses KaTeX for other parsing, so it'd be great to have this is possible. 

Is the font loading still broken?

**Comment by SamSaffron** [comment]
@cycomachead can you do a public topic on meta.discourse.org asking for help getting KaTeX support, explain why it is important and different to MathJax in the post.

At the moment we are running low on staff members working for us that can take this so we may need some help from the community.

**Comment by cycomachead** [comment]
Thanks for the quick response! :)

I started writing a post -- but then realized what I needed won't quite be solved by this. I'm looking to use KaTeX since we use it in other (non-Discourse), but I'm also really interested in being able to easily configure KaTeX. Right now, that involves changing the delimiters (we use $$ and $$$) and eventually I might create a few macros.

So, for now, I found a bit of less clean but easily solution in just dropping Katex in a theme component. Here's a [gist](https://gist.github.com/cycomachead/e388ec65269e2e9adf95d0eae5bfc418) of a very simple version.

**Comment by danielhollas** [comment]
@SamSaffron I was able to fix the fonts by rebuilding KaTeX and putting absolute paths into katex.min.css. Not sure it's the easiest/right thing to do, but it works and we would have the most up-to-date KaTeX. With that change I think this would be ready. I might be able to help with some handholding perhaps? What would be a next step?

---
Steps to build KaTeX:
1. `git clone https://github.com/KaTeX/KaTeX.git && cd KaTeX`
    `git submodule init && git submodule update`

2. Change paths to fonts:
    `sed -ri 's/@font-folder.+$/@font-folder: "\/plugins\/discourse-math\/katex\/fonts";/'  submodules/katex-fonts/fonts.less`

3. Disable TTF fonts:
   `sed -ri 's/@use-ttf.+$/@use-ttf: false;/'  submodules/katex-fonts/fonts.less`

3. Build KaTeX
   `yarn && yarn builld`

4. Copy `katex.min.js` and `katex.min.css` from `dist/` directory

**Comment by danielhollas** [comment]
So KaTeX just released a new version. I did a new custom rebuild according to the steps above to fix font loading and remove TTF fonts. I've rebased this branch on master and added my modifications [in my forked repo](https://github.com/danielhollas/discourse-math/). @SamSaffron I believe with my changes this is ready to be reviewed. Should I open a new PR from my repo? Or is there a better way? Thanks!

**Comment by SamSaffron** [comment]
@danielhollas sure open another PR @majakomel can help review it and get it merged.

**Comment by danielhollas** [comment]
@SamSaffron thanks for merging! This one can be closed. :-)

**Pull request closed without merging**